modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C

```diff
@@ -0,0 +1,505 @@
+      "coefficients for the 'y' component velocity, solved "
+      "via the Navier-Stokes equations.");
+  params.addParam<MooseFunctorName>(
+      "a_w",
+      "For simulations in which the advecting velocities are aux variables, this parameter must be "
+      "supplied when the mesh dimension is greater than 2. It represents the on-diagonal "
+      "coefficients for the 'z' component velocity, solved "
+      "via the Navier-Stokes equations.");
+  return params;
+}
+
+INSFVRhieChowInterpolator::INSFVRhieChowInterpolator(const InputParameters & params)
+  : GeneralUserObject(params),
+    TaggingInterface(this),
+    BlockRestrictable(this),
+    _moose_mesh(UserObject::_subproblem.mesh()),
+    _mesh(_moose_mesh.getMesh()),
+    _dim(_moose_mesh.dimension()),
+    _vel(libMesh::n_threads()),
+    _p(dynamic_cast<INSFVPressureVariable *>(
+        &UserObject::_subproblem.getVariable(0, getParam<VariableName>(NS::pressure)))),
+    _u(dynamic_cast<INSFVVelocityVariable *>(
+        &UserObject::_subproblem.getVariable(0, getParam<VariableName>("u")))),
+    _v(isParamValid("v") ? dynamic_cast<INSFVVelocityVariable *>(
+                               &UserObject::_subproblem.getVariable(0, getParam<VariableName>("v")))
+                         : nullptr),
+    _w(isParamValid("w") ? dynamic_cast<INSFVVelocityVariable *>(
+                               &UserObject::_subproblem.getVariable(0, getParam<VariableName>("w")))
+                         : nullptr),
+    _ps(libMesh::n_threads(), nullptr),
+    _us(libMesh::n_threads(), nullptr),
+    _vs(libMesh::n_threads(), nullptr),
+    _ws(libMesh::n_threads(), nullptr),
+    _sub_ids(blockRestricted() ? blockIDs() : _moose_mesh.meshSubdomains()),
+    _a(_moose_mesh, _sub_ids),
+    _ax(_a, 0),
+    _ay(_a, 1),
+    _az(_a, 2),
+    _sys(*getCheckedPointerParam<SystemBase *>("_sys")),
+    _example(0),
+    _a_data_provided(false)
+{
+  if (!_p)
+    paramError(NS::pressure, "the pressure must be a INSFVPressureVariable.");
+
+  auto fill_container = [this](const auto & name, auto & container)
+  {
+    for (const auto tid : make_range(libMesh::n_threads()))
+    {
+      auto * const var = static_cast<MooseVariableFVReal *>(
+          &UserObject::_subproblem.getVariable(tid, getParam<VariableName>(name)));
+      container[tid] = var;
+    }
+  };
+
+  auto check_blocks = [this](const auto & var)
+  {
+    if (blockIDs() != var.blockIDs())
+      mooseError("Block restriction of interpolator user object '",
+                 this->name(),
+                 "' doesn't match the block restriction of variable '",
+                 var.name(),
+                 "'");
+  };
+
+  fill_container(NS::pressure, _ps);
+  check_blocks(*_p);
+
+  if (!_u)
+    paramError("u", "the u velocity must be an INSFVVelocityVariable.");
+  fill_container("u", _us);
+  check_blocks(*_u);
+  _var_numbers.push_back(_u->number());
```

Comment:
```suggestion
  fillARead();
  if (_velocity_interp_method != Moose::FV::InterpMethod::Average && !_a_data_provided)
  _var_numbers.push_back(_u->number());
```
I know this doesnt work right now, but I think we need both of these conditions (the latter more than the first one).
- we dont need to gather the RC data if we're doing average interpolation. And we dont need to be checking the momentum kernels to be be momentum ROs. Since I m ok with disallowing average for the velocity/pressure solve, I m ok with not implementing this. However, we still need a workaround when we do scalar with average velocity interpolation, and therefore do not bother providing ax and ay.
- if u is an auxvariable, this will hold the wrong variable number, triggering the check on the wrong kernel
- if this variable number vector is empty, everything we dont want just gets skipped naturally

Reply:
Ok I am 100% on not allowing average for this.

Reply:
then we just need to be able to skip the momentum residual object check when providing a_x a_y directly.
I m not sure how you got that to work in your test, I am hitting this error in my MSR advection problem

Reply:
yup execute() is skipped but initialSetup is still being ran.
As you want for the fix. Could just return in initalSetup

Reply:
You must have residual objects for your velocities in this failing input? That's the only possible way you could be getting an error

Reply:
if you tell me what the input is, then I can investigate

Reply:
sending on slack

Reply:
> (since rn it s being filled with the 'wrong' numbers)

It's being filled exactly with the velocity variable's numbers :smile: I would hardly call them wrong

Reply:
oh i do, I have the boundary conditions. I guess those are no longer needed

Reply:
they are the variable numbers in the wrong system I think.
It s using those numbers to gather kernels for the non linear variables, when they are auxiliary